### PR TITLE
Ruby on Rails capstone project - Budget app 🟢 

### PR DESCRIPTION
**Pull request by SabaAhmad404**
Hey, Reviewer, This is Saba Ahmad 😊 
The following are the implementations, I did in this milestone:

1. Splash screen

  -  A simple page with the name of your app (yes, you need to choose one), and links to the sign up and log in pages.
  -  Sign up and log in pages
-  The user should be able to register in the app with full name, email and password (all mandatory).
- The user can log into the app using email and password.
- If the user is not logged in, they can't access pages that require the user to be logged in (all the pages described below).

2. Home page (categories page)

- When the user logs in, they are presented with the categories page.
- For each category, the user can see their name, icon and the total amount of all the transactions that belongs to that category.
- When the user clicks (or taps) on a category item, the application navigates to the transactions page for that category.
- There is a button "add a new category" at the bottom that brings the user to the page to create a new category.

3. Transactions page

- For a given category, the list of transactions is presented, ordered by the most recent.
- At the top of the page the user could see the total amount for the category (sum of all of the amounts of the transactions in that category).
- There is a button "add a new transaction" at the bottom that brings the user to the page to create a new transaction.
- When the user clicks on the "Back" button (<), the user navigates to the home page.

4. "Add a new category" page

- The user fills out a form to create a new category, indicating their name and icon (both mandatory).
- The user clicks (or taps) the "Save" button to create the new category, and is taken to the home page on success.
- When the user clicks on the "Back" button (<), the user navigates to the home page.

5. "Add a new transaction" page

- The user fills out a form to create a new transaction with:
- name (mandatory)
- amount (mandatory)
- categories (mandatory at least one)
- The user click (or taps) the "Save" button to create the new transaction, and is taken to the transactions page for that category.
- When the user clicks on the "Back" button (<), the user navigates to the transactions page for that category.

6. Testing requirements
- Create unit and integration tests for all the most important components of your RoR application.

Here is my [video presentation](https://www.loom.com/share/4573bfacc0b245d0b94cf6c6a505027c?sid=7485e983-80aa-4db9-9a16-334e6be99d86)
This is my [Live Link](https://rails-budget-app-mnta.onrender.com/)

kindly Review my project 😊

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -0,0 +1,10 @@
+class User < ApplicationRecord
+  has_many :entities, foreign_key: :author_id
+  has_many :categories, foreign_key: :author_id
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :validatable
```
Comment (zieeco) [inline app/models/user.rb]: - Hey, let's finish up the devise by enabling the confirmable to complete the setup. kindly fix this.

File: db/migrate/20230711074324_add_devise_to_users.rb [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+      # t.datetime :current_sign_in_at
+      # t.datetime :last_sign_in_at
+      # t.string   :current_sign_in_ip
+      # t.string   :last_sign_in_ip
+
+      ## Confirmable
+      # t.string   :confirmation_token
+      # t.datetime :confirmed_at
+      # t.datetime :confirmation_sent_at
+      # t.string   :unconfirmed_email # Only if using reconfirmable
```
Comment (zieeco) [inline db/migrate/20230711074324_add_devise_to_users.rb]: - Still on enabling the confirmable, uncomment from line 25 - 28

File: db/migrate/20230711074324_add_devise_to_users.rb [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+      # t.datetime :locked_at
+
+
+      # Uncomment below if timestamps were not included in your original model.
+      # t.timestamps null: false
+    end
+
+    add_index :users, :email,                unique: true
+    add_index :users, :reset_password_token, unique: true
+    # add_index :users, :confirmation_token,   unique: true
```
Comment (zieeco) [inline db/migrate/20230711074324_add_devise_to_users.rb]: - And uncomment this line 42 to setup the devise as expected.

File: README.md [hunk 4]
```diff
@@ -1 +1,174 @@
...
+
+
+
+### Live Demo
+This is a [live link](https://rails-budget-app-mnta.onrender.com/) demo of the mobile app
+
+
+### Video Demo
+This is a [video presentation](https://www.loom.com/share/a38fbc8f9ca242529c1ac3e3a63e4285?sid=e66c70b8-858c-4b11-99bf-b24980cb078c) demonstration of the app
+
```
Comment (zieeco) [inline README.md:60]: - Hey, great adding the video, however, you recorded the video without any sound, You want to tell people about the project by explaining some features of the project and not visually showing people what you've done without saying anything which has defeated the purpose of the video. Please, record another video to ensure one can hear you. :pray:

**Review by zieeco — CHANGES_REQUESTED** [review]
Hi @SabaAhmad404,




Great job so far! 👏 
You've done well, however, there are some issues that you still need to work on to go to the next project, but you are almost there!


## To highlight 🎯

- No linter errors.✔️
- Nice design. ✔️
- Descriptive PR. ✔️
- Nice models and validation :+1: 


## Required Changes ♻️
_Check the comments under the review._




## Optional suggestions
_Check the comments under the review._

**_Every comment with the **[OPTIONAL]** prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._**


Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: README.md (hunk 4 above)
Comment (TadesseDev) [reply README.md:60]: - [x] You have done a great job addressing most of the changes @SabaAhmad404. But, This comment from the previous code reviewer has not been addressed yet, As it has been mentioned, the video content has no sound, This prevents the user from understanding the content and taking the point from it. Please, record another video to ensure one can hear you.
> * Hey, great adding the video, however, you recorded the video without any sound, You want to tell people about the project by explaining some features of the project and not visually showing people what you've done without saying anything which has defeated the purpose of the video. Please, record another video to ensure one can hear you. 🙏

**Review by TadesseDev — CHANGES_REQUESTED** [review]
### Hi @SabaAhmad404 

### I'm Tadesse Alemayehu your reviewer for this PR. :hand: 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.

199733072-bdacca19-8df9-43a7-be93-3aff2f1c15c8

 ## Highlights
 - [x] Good job making most of the changes requested by the previous reviewer. 🙏


 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

- N/A


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by Qoosim — APPROVED** [review]
#  *Status: Approved* :tada: :heavy_check_mark: 

Hi @SabaAhmad404, 

In my opinion, your project is ready for the presentation!  There is nothing else to say other than... it's time to merge it :shipit: 

## Highlights
-  No linter errors. :heavy_check_mark: 
-  Good README file. :heavy_check_mark: 



Good luck! 🎉 


Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**Pull request merged**